selftest.py

```diff
@@ -0,0 +1,136 @@
+"""Define the interfaces used by ExternalIntegration self-tests.
+"""
+from nose.tools import set_trace
+from util.http import IntegrationException
+import datetime
+
+
+class SelfTestResult(object):
+    """The result of running a single self-test.
+
+    HasSelfTest.run_self_tests() returns a list of these
+    """
+
+    def __init__(self, name):
+        # Name of the test.
+        self.name = name
+
+        # Set to True when the test runs without raising an exception.
+        self.success = False
+
+        # The exception raised, if any.
+        self.exception = None
+
+        # The return value of the test method, assuming it ran to
+        # completion.
+        self.result = None
+
+        # Start time of the test.
+        self.start = datetime.datetime.utcnow()
+
+        # End time of the test.
+        self.end = None
+
+    def __repr__(self):
+        if self.exception:
+            if (isinstance(self.exception, IntegrationException)
+                and self.exception.debug_message):
+                exception = " exception=%r debug=%r" % (
+                    self.exception.message, self.exception.debug_message
+                )
+            else:
+                exception = " exception=%r" % self.exception
+        else:
+            exception = ""
+        return "<SelfTestResult: name=%r timing=%.2fsec success=%r%s result=%r>" % (
+            self.name, (self.end-self.start).total_seconds(), self.success,
+            exception, self.result
+        )
+
+
+class HasSelfTests(object):
+    """An object capable of verifying its own setup by running a
+    series of self-tests.
+    """
+
+    @classmethod
+    def run_self_tests(cls, _db, constructor_method=None, *args, **kwargs):
+        """Instantiate this class and call _run_self_tests on it.
+
+        :param _db: A database connection. Will be passed into
+        _run_self_tests as well as the first argument of the constructor
+        method.
+
+        :param constructor_method: Method to use to instantiate the
+        class, if different from the default constructor.
```

Comment:
We need this because some integration classes want to be instantiated with `from_config` or another similar method, and some want their normal constructor to be used.